libwebrtc-hw/vpl_video_encoder.cpp

```diff
@@ -0,0 +1,512 @@
+#include <modules/video_coding/include/video_codec_interface.h>
+#include <modules/video_coding/include/video_error_codes.h>
+#include <rtc_base/logging.h>
+#include <rtc_base/synchronization/mutex.h>
+
+// Intel VPL
+#include <fcntl.h>
+#include <mfxdispatcher.h>
+#include <mfxvideo++.h>
+#include <mfxvideo.h>
+#include <mfxvp8.h>
+#include "va/va.h"
+#include "va/va_drm.h"
+// libyuv
+#include <libyuv.h>
+
+#include "vpl_utils.h"
+
+namespace {
+
+constexpr float MIN_ADJUSTED_BITRATE_PERCENTAGE = 0.5;
+constexpr float MAX_ADJUSTED_BITRATE_PERCENTAGE = 0.95;
+
+}  // namespace
+
+namespace any_vpl {
+
+VplVideoEncoder::VplVideoEncoder(webrtc::VideoCodecType codec)
+    : codec_(ToMfxCodec(codec)), bitrateAdjuster_(MIN_ADJUSTED_BITRATE_PERCENTAGE, MAX_ADJUSTED_BITRATE_PERCENTAGE) {}
+
+VplVideoEncoder::~VplVideoEncoder() {
+  Release();
+}
+
+mfxStatus VplVideoEncoder::ExecQuery(mfxVideoParam& param) {
+  mfxVideoParam queryParam;
+  memcpy(&queryParam, &param, sizeof(param));
+  mfxStatus mfxSts = encoder_->Query(&queryParam, &queryParam);
+
+  if (mfxSts >= 0) {
+#define PrintParamInfo(NAME) \
+  if (param.NAME != queryParam.NAME) RTC_LOG(LS_WARNING) << "param " << #NAME << " old=" << param.NAME << " new=" << queryParam.NAME
+    PrintParamInfo(mfx.LowPower);
+    PrintParamInfo(mfx.BRCParamMultiplier);
+    PrintParamInfo(mfx.FrameInfo.FrameRateExtN);
+    PrintParamInfo(mfx.FrameInfo.FrameRateExtD);
+    PrintParamInfo(mfx.FrameInfo.FourCC);
+    PrintParamInfo(mfx.FrameInfo.ChromaFormat);
+    PrintParamInfo(mfx.FrameInfo.PicStruct);
+    PrintParamInfo(mfx.FrameInfo.CropX);
+    PrintParamInfo(mfx.FrameInfo.CropY);
+    PrintParamInfo(mfx.FrameInfo.CropW);
+    PrintParamInfo(mfx.FrameInfo.CropH);
+    PrintParamInfo(mfx.FrameInfo.Width);
+    PrintParamInfo(mfx.FrameInfo.Height);
+    PrintParamInfo(mfx.CodecId);
+    PrintParamInfo(mfx.CodecProfile);
+    PrintParamInfo(mfx.CodecLevel);
+    PrintParamInfo(mfx.GopPicSize);
+    PrintParamInfo(mfx.GopRefDist);
+    PrintParamInfo(mfx.GopOptFlag);
+    PrintParamInfo(mfx.IdrInterval);
+    PrintParamInfo(mfx.TargetUsage);
+    PrintParamInfo(mfx.RateControlMethod);
+    PrintParamInfo(mfx.InitialDelayInKB);
+    PrintParamInfo(mfx.TargetKbps);
+    PrintParamInfo(mfx.MaxKbps);
+    PrintParamInfo(mfx.BufferSizeInKB);
+    PrintParamInfo(mfx.NumSlice);
+    PrintParamInfo(mfx.NumRefFrame);
+    PrintParamInfo(mfx.EncodedOrder);
+    PrintParamInfo(mfx.DecodedOrder);
+    PrintParamInfo(mfx.ExtendedPicStruct);
+    PrintParamInfo(mfx.TimeStampCalc);
+    PrintParamInfo(mfx.SliceGroupsPresent);
+    PrintParamInfo(mfx.MaxDecFrameBuffering);
+    PrintParamInfo(mfx.EnableReallocRequest);
+    PrintParamInfo(AsyncDepth);
+    PrintParamInfo(IOPattern);
+
+    memcpy(&param, &queryParam, sizeof(param));
+  }
+  return mfxSts;
+}
+
+mfxStatus VplVideoEncoder::ExecQueries(mfxVideoParam& param, ExtBuffer& ext) {
+  mfxStatus mfxSts = MFX_ERR_NONE;
+
+  memset(&param, 0, sizeof(param));
```

Comment:
why not default constructor?

```suggestion
  param = mfxVideoParam();
```

Reply:
I think that the default constructor doesn't guarantee to initialize the primitive types to zero, so we might have integers with garbage